File: apps/web/src/composables/market-data/candleSessions.ts
export type CandleSession = "regular" | "extended" | "overnight";

export const CANDLE_SESSION_ORDER: readonly CandleSession[] = [
  "regular",
  "extended",
  "overnight",
];

export const CANDLE_SESSION_LABELS: Record<CandleSession, string> = {
  regular: "盘中",
  extended: "盘前后",
  overnight: "夜盘",
};

export function normalizeCandleSessions(
  values: readonly (string | null | undefined)[] | null | undefined,
): CandleSession[] {
  const requested = new Set(values?.flatMap((value) =>
    String(value ?? "")
      .split(",")
      .map((token) => token.trim().toLowerCase()),
  ));
  return CANDLE_SESSION_ORDER.filter((session) => requested.has(session));
}

export function intersectCandleSessions(
  selected: readonly CandleSession[],
  available: readonly CandleSession[],
): CandleSession[] {
  const availableSet = new Set(available);
  return CANDLE_SESSION_ORDER.filter(
    (session) => availableSet.has(session) && selected.includes(session),
  );
}

export function summarizeCandleSessions(
  sessions: readonly CandleSession[],
): string {
  const normalized = normalizeCandleSessions(sessions);
  if (normalized.length === CANDLE_SESSION_ORDER.length) return "全天";
  if (normalized.length === 1) return CANDLE_SESSION_LABELS[normalized[0]!];
  if (normalized.length === 2 && normalized[0] === "regular" && normalized[1] === "extended") {
    return "盘中+盘前后";
  }
  return normalized.map((session) => CANDLE_SESSION_LABELS[session]).join("+");
}

export function candleSessionForValue(
  value: string | null | undefined,
): CandleSession | null {
  const normalized = value?.trim().toLowerCase();
  if (normalized === "regular" || normalized === "rth") return "regular";
  if (normalized === "extended" || normalized === "pre" || normalized === "after" || normalized === "eth") return "extended";
  if (normalized === "overnight") return "overnight";
  return null;
}

export function candleMatchesSessions(
  value: string | null | undefined,
  selected: readonly CandleSession[],
): boolean {
  const session = candleSessionForValue(value) ?? "regular";
  return selected.includes(session);
}

File: internal/marketdata/candle_sessions.go
package marketdata

import (
	"fmt"
	"slices"
	"strings"
)

type CandleSession string

const (
	CandleSessionRegular   CandleSession = "regular"
	CandleSessionExtended  CandleSession = "extended"
	CandleSessionOvernight CandleSession = "overnight"
)

var candleSessionOrder = []CandleSession{
	CandleSessionRegular,
	CandleSessionExtended,
	CandleSessionOvernight,
}

// HistoricalCandlesQuery keeps every request dimension together so provider
// adapters cannot accidentally omit session scope during pagination.
type HistoricalCandlesQuery struct {
	Market            string
	Symbol            string
	Period            string
	Limit             int
	FromTime          string
	ToTime            string
	// BeforeTime is an exclusive historical-page cursor. It must remain
	// distinct from ToTime because upstream providers can use coarser time
	// precision than the public RFC3339 value.
	BeforeTime        string
	Sessions          []CandleSession
	SessionsSpecified bool
}

func ParseCandleSessions(values []string) ([]CandleSession, error) {
	seen := make(map[CandleSession]struct{}, len(candleSessionOrder))
	provided := false
	for _, value := range values {
		for _, token := range strings.Split(value, ",") {
			provided = true
			token = strings.TrimSpace(token)
			if token == "" {
				continue
			}
			session := CandleSession(strings.ToLower(token))
			if !slices.Contains(candleSessionOrder, session) {
				return nil, fmt.Errorf("%w: %q", ErrInvalidCandleSessions, token)
			}
			seen[session] = struct{}{}
		}
	}
	if provided && len(seen) == 0 {
		return nil, fmt.Errorf("%w: at least one session is required", ErrInvalidCandleSessions)
	}
	return orderedCandleSessions(seen), nil
}

func ResolveCandleSessions(
	requested []CandleSession,
	specified bool,
	available []CandleSession,
) ([]CandleSession, error) {
	availableSet := make(map[CandleSession]struct{}, len(available))
	for _, session := range available {
		availableSet[session] = struct{}{}
	}
	if !specified {
		return orderedCandleSessions(availableSet), nil
	}
	if len(requested) == 0 {
		return nil, fmt.Errorf("%w: at least one session is required", ErrInvalidCandleSessions)
	}
	requestedSet := make(map[CandleSession]struct{}, len(requested))
	for _, session := range requested {
		if _, ok := availableSet[session]; !ok {
			return nil, fmt.Errorf("%w: session %q is unsupported", ErrInvalidCandleSessions, session)
		}
		requestedSet[session] = struct{}{}
	}
	return orderedCandleSessions(requestedSet), nil
}

func CandleSessionStrings(sessions []CandleSession) []string {
	values := make([]string, 0, len(sessions))
	for _, session := range sessions {
		values = append(values, string(session))
	}
	return values
}

func ContainsCandleSession(sessions []CandleSession, wanted CandleSession) bool {
	return slices.Contains(sessions, wanted)
}

func CandleSessionForLabel(label string) CandleSession {
	switch strings.ToLower(strings.TrimSpace(label)) {
	case "regular":
		return CandleSessionRegular
	case "pre", "after", "extended":
		return CandleSessionExtended
	case "overnight":
		return CandleSessionOvernight
	default:
		return ""
	}
}

func FilterCandlesBySessions(
	candles []map[string]any,
	sessions []CandleSession,
) []map[string]any {
	if len(sessions) == 0 {
		return candles
	}
	filtered := make([]map[string]any, 0, len(candles))
	for _, candle := range candles {
		label, _ := candle["session"].(string)
		group := CandleSessionForLabel(label)
		if group == "" {
			group = CandleSessionRegular
		}
		if ContainsCandleSession(sessions, group) {
			filtered = append(filtered, candle)
		}
	}
	return filtered
}

func orderedCandleSessions(set map[CandleSession]struct{}) []CandleSession {
	result := make([]CandleSession, 0, len(set))
	for _, session := range candleSessionOrder {
		if _, ok := set[session]; ok {
			result = append(result, session)
		}
	}
	return result
}
